File: NextApp/Android/feature/tools/src/main/kotlin/jp/komehyappyo/member/next/feature/tools/MeetingMinutesFeatureModel.kt
package jp.komehyappyo.member.next.feature.tools

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import jp.komehyappyo.member.next.core.data.LocalMeetingRecordingStore
import jp.komehyappyo.member.next.core.data.MeetingMinutesRepository
import jp.komehyappyo.member.next.core.model.MeetingMinutes
import jp.komehyappyo.member.next.core.model.MeetingRecordingDraft
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import java.io.File
import java.time.Instant
import java.util.UUID

data class MeetingMinutesUiState(
    val minutes: List<MeetingMinutes> = emptyList(),
    val draft: MeetingRecordingDraft? = null,
    val isRecording: Boolean = false,
    val isTranscribing: Boolean = false,
    val elapsedSeconds: Int = 0,
    val liveTranscript: String = "",
    val notice: String? = null,
    val errorMessage: String? = null,
)

class MeetingMinutesFeatureModel(
    private val repository: MeetingMinutesRepository,
    private val recordingStore: LocalMeetingRecordingStore,
    private val recorder: MeetingRecordingService,
) : ViewModel() {
    private val mutableState = MutableStateFlow(
        runCatching { recordingStore.load() }.getOrNull().let { draft ->
            MeetingMinutesUiState(
                draft = draft,
                elapsedSeconds = draft?.recordingDurationSeconds ?: 0,
                liveTranscript = draft?.transcriptText.orEmpty(),
            )
        },
    )
    val state: StateFlow<MeetingMinutesUiState> = mutableState
    private var timerJob: Job? = null

    init {
        recorder.onTranscriptChanged = { value ->
            mutableState.update { it.copy(liveTranscript = value) }
        }
        recorder.onTranscriptionStateChanged = { value ->
            mutableState.update {
                it.copy(
                    isTranscribing = value,
                    liveTranscript = recorder.transcript,
                    notice = recorder.transcriptionNotice,
                )
            }
            if (!value) persistDraft()
        }
        recorder.onInterrupted = {
            stopRecording(interrupted = true)
        }
        viewModelScope.launch {
            repository.observeAll()
                .catch { error ->
                    mutableState.update { it.copy(errorMessage = error.localizedMessage) }
                }
                .collect { values -> mutableState.update { it.copy(minutes = values) } }
        }
    }

    fun startRecording(): Result<Unit> = runCatching {
        val id = UUID.randomUUID()
        val file = recordingStore.newTemporaryAudioFile(id)
        val draft = MeetingRecordingDraft(id = id, audioFileLocalPath = file.absolutePath)
        recordingStore.save(draft)
        recorder.start(file)
        mutableState.update {
            it.copy(
                draft = draft,
                isRecording = true,
                elapsedSeconds = 0,
                liveTranscript = "",
                notice = recorder.transcriptionNotice,
                errorMessage = null,
            )
        }
        timerJob = viewModelScope.launch {
            while (isActive) {
                delay(1_000)
                mutableState.update { current ->
                    current.copy(
                        elapsedSeconds = current.elapsedSeconds + 1,
                        liveTranscript = recorder.transcript,
                        notice = recorder.transcriptionNotice,
                    )
                }
                if (mutableState.value.elapsedSeconds % 10 == 0) persistDraft()
            }
        }
    }.onFailure { error ->
        mutableState.update { it.copy(errorMessage = error.localizedMessage) }
    }

    fun stopRecording(interrupted: Boolean = false) {
        val text = recorder.stopAndTranscribe()
        timerJob?.cancel()
        timerJob = null
        mutableState.update {
            it.copy(
                isRecording = false,
                isTranscribing = recorder.isTranscribing,
                liveTranscript = text,
                notice = if (interrupted) {
                    "録音が中断されました。録音を保存し、文字起こしを試みています。"
                } else {
                    recorder.transcriptionNotice
                },
            )
        }
        persistDraft()
    }

    fun retryTranscription() {
        val draft = mutableState.value.draft ?: return
        val file = File(draft.audioFileLocalPath)
        if (!file.exists()) {
            reportError("文字起こしする録音ファイルが見つかりません。")
            return
        }
        mutableState.update {
            it.copy(
                isTranscribing = true,
                liveTranscript = "",
                notice = "保存した録音の文字起こしを再実行します。",
                errorMessage = null,
            )
        }
        recorder.transcribe(file)
    }

    fun saveDraft(title: String, transcript: String, onComplete: (Result<Unit>) -> Unit) {
        viewModelScope.launch {
            runCatching {
                val draft = requireNotNull(mutableState.value.draft)
                val now = Instant.now()
                val duration = maxOf(
                    draft.recordingDurationSeconds,
                    mutableState.value.elapsedSeconds,
                )
                val destination = recordingStore.moveAudio(
                    File(draft.audioFileLocalPath),
                    draft.id,
                )
                repository.save(
                    MeetingMinutes(
                        id = draft.id,
                        userId = draft.userId,
                        title = title,
                        recordingStartAt = draft.startedAt,
                        recordingEndAt = draft.startedAt.plusSeconds(duration.toLong()),
                        recordingDurationSeconds = duration,
                        audioFileLocalPath = destination.absolutePath,
                        transcriptText = transcript,
                        createdAt = draft.startedAt,
                        updatedAt = now,
                    ).validated(now),
                )
                recordingStore.delete()
                mutableState.update {
                    it.copy(
                        draft = null,
                        isTranscribing = false,
                        liveTranscript = "",
                        elapsedSeconds = 0,
                    )
                }
            }.also(onComplete)
        }
    }

    fun update(minutes: MeetingMinutes, onComplete: (Result<Unit>) -> Unit) {
        viewModelScope.launch {
            runCatching { repository.save(minutes.copy(updatedAt = Instant.now())) }
                .also(onComplete)
        }
    }

    fun delete(minutes: MeetingMinutes) {
        viewModelScope.launch {
            runCatching { repository.delete(minutes.id) }
                .onFailure { error ->
                    mutableState.update { it.copy(errorMessage = error.localizedMessage) }
                }
        }
    }

    fun discardDraft() {
        mutableState.value.draft?.let(recordingStore::discard)
        mutableState.update {
            it.copy(
                draft = null,
                isTranscribing = false,
                liveTranscript = "",
                elapsedSeconds = 0,
            )
        }
    }

    fun clearError() {
        mutableState.update { it.copy(errorMessage = null) }
    }

    fun reportError(message: String) {
        mutableState.update { it.copy(errorMessage = message) }
    }

    private fun persistDraft() {
        val current = mutableState.value
        val draft = current.draft ?: return
        val updated = draft.copy(
            transcriptText = current.liveTranscript,
            recordingDurationSeconds = current.elapsedSeconds,
            updatedAt = Instant.now(),
        )
        runCatching { recordingStore.save(updated) }
            .onSuccess { mutableState.update { it.copy(draft = updated) } }
            .onFailure { error ->
                mutableState.update { it.copy(errorMessage = error.localizedMessage) }
            }
    }

    override fun onCleared() {
        if (recorder.isRecording) {
            timerJob?.cancel()
            persistDraft()
        }
        recorder.release()
    }

    class Factory(
        private val repository: MeetingMinutesRepository,
        private val recordingStore: LocalMeetingRecordingStore,
        private val recorder: MeetingRecordingService,
    ) : ViewModelProvider.Factory {
        @Suppress("UNCHECKED_CAST")
        override fun <T : ViewModel> create(modelClass: Class<T>): T =
            MeetingMinutesFeatureModel(repository, recordingStore, recorder) as T
    }
}

File: NextApp/iOS/OrgPortalNext/Feature/Tools/MeetingMinutesFeatureModel.swift
import AVFAudio
import DataLayer
import Foundation
import Model

@MainActor
public final class MeetingMinutesFeatureModel: ObservableObject {
    @Published public private(set) var minutes: [MeetingMinutes] = []
    @Published public private(set) var draft: MeetingRecordingDraft?
    @Published public private(set) var isRecording = false
    @Published public private(set) var elapsedSeconds = 0
    @Published public private(set) var liveTranscript = ""
    @Published public private(set) var notice: String?
    @Published public private(set) var errorMessage: String?

    public let recorder: MeetingRecordingService
    private let repository: MeetingMinutesRepository
    private let recordingStore: LocalMeetingRecordingStore
    private var timer: Timer?

    public init(
        repository: MeetingMinutesRepository,
        recordingStore: LocalMeetingRecordingStore,
        recorder: MeetingRecordingService = MeetingRecordingService()
    ) {
        self.repository = repository
        self.recordingStore = recordingStore
        self.recorder = recorder
        recorder.onInterruption = { [weak self] in
            self?.finishInterruptedRecording()
        }
    }

    public func load() async {
        do {
            minutes = try await repository.fetchAll()
            let restoredDraft = try recordingStore.load()
            draft = restoredDraft
            elapsedSeconds = restoredDraft?.recordingDurationSeconds ?? 0
            liveTranscript = restoredDraft?.transcriptText ?? ""
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    public func startRecording() async -> Bool {
        let permissions = await MeetingRecordingService.requestPermissions()
        guard permissions.microphoneGranted else {
            errorMessage =
                "録音にはマイクの許可が必要です。設定アプリから許可してください。"
            return false
        }
        do {
            let id = UUID()
            let audioURL = try recordingStore.newTemporaryAudioURL(id: id)
            let value = MeetingRecordingDraft(
                id: id,
                audioFileLocalPath: audioURL.path
            )
            try recordingStore.save(value)
            draft = value
            try recorder.start(
                at: audioURL,
                transcriptionEnabled: permissions.speechRecognitionGranted
            )
            isRecording = true
            elapsedSeconds = 0
            liveTranscript = ""
            notice = recorder.transcriptionNotice
            timer = Timer.scheduledTimer(withTimeInterval: 1, repeats: true) {
                [weak self] _ in
                Task { @MainActor in self?.tick() }
            }
            return true
        } catch {
            errorMessage = "録音を開始できませんでした。\(error.localizedDescription)"
            return false
        }
    }

    public func stopRecording() {
        liveTranscript = recorder.stop()
        isRecording = false
        timer?.invalidate()
        timer = nil
        persistDraft()
    }

    public func saveDraft(title: String, transcript: String) async throws {
        guard let draft else { return }
        let now = Date()
        let duration = max(draft.recordingDurationSeconds, elapsedSeconds)
        let destination = try recordingStore.moveAudio(
            from: URL(fileURLWithPath: draft.audioFileLocalPath),
            minutesId: draft.id
        )
        let value = try MeetingMinutes(
            id: draft.id,
            userId: draft.userId,
            title: title,
            recordingStartAt: draft.startedAt,
            recordingEndAt: draft.startedAt.addingTimeInterval(TimeInterval(duration)),
            recordingDurationSeconds: duration,
            audioFileLocalPath: destination.path,
            transcriptText: transcript,
            createdAt: draft.startedAt,
            updatedAt: now
        ).validated(now: now)
        try await repository.save(value)
        try recordingStore.delete()
        self.draft = nil
        liveTranscript = ""
        elapsedSeconds = 0
        await load()
    }

    public func update(_ value: MeetingMinutes) async throws {
        try await repository.save(value)
        await load()
    }

    public func delete(_ value: MeetingMinutes) async {
        do {
            try await repository.delete(id: value.id)
            await load()
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    public func discardDraft() {
        guard let draft else { return }
        do {
            try recordingStore.discard(draft)
            self.draft = nil
            liveTranscript = ""
            elapsedSeconds = 0
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    public func report(_ error: Error) {
        errorMessage = error.localizedDescription
    }

    public func clearError() {
        errorMessage = nil
    }

    private func tick() {
        elapsedSeconds += 1
        liveTranscript = recorder.transcript
        notice = recorder.transcriptionNotice
        if elapsedSeconds.isMultiple(of: 10) {
            persistDraft()
        }
    }

    private func persistDraft() {
        guard var draft else { return }
        draft.transcriptText = liveTranscript
        draft.recordingDurationSeconds = elapsedSeconds
        draft.updatedAt = .now
        do {
            try recordingStore.save(draft)
            self.draft = draft
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    private func finishInterruptedRecording() {
        isRecording = false
        timer?.invalidate()
        timer = nil
        liveTranscript = recorder.transcript
        persistDraft()
        notice = "録音が中断されました。未保存の録音として復旧できます。"
    }
}
